raft: test TestCannotCommitWithoutNewTermEntry w/ & wo/ store liveness

iskettaneh opened the pull request:
This commit changes TestCannotCommitWithoutNewTermEntry to allow testing it when store liveness is both enabled and disabled.

In order to achieve that, it adds a mockStoreLiveness for raft unit tests. It acts like AlwaysLive store liveness by default, but it could be to change the supportFor/SupportFrom between any two nodes, also, it allows to bump the support epochs.

References: #132241
    
Release note: None

iskettaneh committed:
raft: create MockStoreLiveness for unit tests

This commit creates MockStoreLiveness to be used in Raft unit tests.
Initially, it acts like the `AlwaysLive` store liveness, but it has
methods that can alter its behavior to withdraw/grant support to other
stores, and bump the support epoch.

Epic: None

Release note: None

iskettaneh committed:
raft: add withStoreLiveness test modifier

This commit adds the test modifier withStoreLiveness which allows us
to provide any StoreLiveness implementation to the test.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/132242)
<!-- Reviewable:end -->

arulajmani reviewed (commented):
> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132242)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/raft/raft_test.go` line 4170 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O8qgDhrD0kN0bXWotvv:-O8qgDhrD0kN0bXWotvw:b-vng6nf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5d66fa455896d97e2d873b16847913f06199719/pkg/raft/raft_test.go#L4170)):*
> ```Go
> // bump the support epoch.
> type mockStoreLiveness struct {
> 	supportEpoch pb.Epoch
> ```

I worry we'll need something more general than this if we're to use this in all tests. This version of StoreLiveness is less expressive than the one we have in the datadriven tests, and I'm struggling with how restrictive that is over in https://github.com/cockroachdb/cockroach/pull/132108. 

The reason I'm struggling there is that if you return `hlc.MaxTimestamp` in `SupportFrom` during the `LeadSupportUntil` calculation, a leader will never de-fortify (because we don't de-fortify until the `maxLeadSupportUntil` is in the past). This means that all tests where the leader gets demoted to a learner and steps down while switching configurations (read: the cases that require de-fortification to elect a new leader) will not work with this simplified version of StoreLiveness. 

Separately, does this simplified version work for all tests that we'll need to change here? Are there ever test cases that need a per-store support epoch? Or is the plan to expand this if/when we find cases like that?


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132242)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/raft/raft_test.go` line 4170 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O8qgDhrD0kN0bXWotvv:-O8rKx9NEiL1GMYWpSMX:befs8t7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5d66fa455896d97e2d873b16847913f06199719/pkg/raft/raft_test.go#L4170)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I worry we'll need something more general than this if we're to use this in all tests. This version of StoreLiveness is less expressive than the one we have in the datadriven tests, and I'm struggling with how restrictive that is over in https://github.com/cockroachdb/cockroach/pull/132108. 

The reason I'm struggling there is that if you return `hlc.MaxTimestamp` in `SupportFrom` during the `LeadSupportUntil` calculation, a leader will never de-fortify (because we don't de-fortify until the `maxLeadSupportUntil` is in the past). This means that all tests where the leader gets demoted to a learner and steps down while switching configurations (read: the cases that require de-fortification to elect a new leader) will not work with this simplified version of StoreLiveness. 

Separately, does this simplified version work for all tests that we'll need to change here? Are there ever test cases that need a per-store support epoch? Or is the plan to expand this if/when we find cases like that?
</blockquote></details>

For the tests that we need to change, I prepared this [document to list them all](https://docs.google.com/document/d/1cSbrdyF_3Tq_ZRsW-UTvxwRB4VylHXL2xEUPovkf3QA/edit).
It seems that all the cases can be fixed by either withdrawing liveness support, or bumping the support epoch.


Regarding the tests that require defortification. Do we have such tests as unit tests? Or do you mean that we will have them in the future?
One idea that comes to my mind if we ever need this is to utilize the fact that we use `SupportExpired()` after calling `LeadSupportUntil()`. So we change the behaviour of `SupportExpired()` in the unit test mock to make it return true when we need that


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132242)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/raft/raft_test.go` line 4170 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O8qgDhrD0kN0bXWotvv:-O8rrKku2aOjp_diN7Tt:b-v6c8k2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5d66fa455896d97e2d873b16847913f06199719/pkg/raft/raft_test.go#L4170)):*
> For the tests that we need to change, I prepared this [document to list them all](https://docs.google.com/document/d/1cSbrdyF_3Tq_ZRsW-UTvxwRB4VylHXL2xEUPovkf3QA/edit).

Nice, thanks for doing this!

> Regarding the tests that require defortification. Do we have such tests as unit tests? Or do you mean that we will have them in the future?

I think tests that require the leader to step down because of `CheckQuorum` (from your doc) will fall into this category as well. The leader will not step down because of `CheckQuorum` until its `MaxLeadSupportUntil` is in the past. With `hlc.MaxTimestamp`, that'll never be true.

We can continue this discussion in the pod meeting later today.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> Reviewed 2 of 2 files at r4, 1 of 1 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132242)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/raft/raft_test.go` line 4165 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O8w1JC3703HUVewMK6w:-O8w1JC3703HUVewMK6x:bgbq27f) ([raw file](https://github.com/cockroachdb/cockroach/blob/96035db1e59585afec494eae70c27df7d2bc9d18/pkg/raft/raft_test.go#L4165)):*
> ```Go
> }
> 
> // mockStoreLiveness is a simple mock implementation of StoreLiveness that
> ```

nits:

s/simple//g

"It initially treats.."

"bump the supported epoch"
___
*[`pkg/raft/raft_test.go` line 716 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O8w5gMM8iQDM1kauCsc:-O8w5gMM8iQDM1kauCsd:b-kt6qg7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3017a723a1fd9b0caac216dc7c460a5591b26770/pkg/raft/raft_test.go#L716)):*
> ```Go
> 	tt.maybeWithdrawSupportAllPeers()
> 
> 	tt.send(pb.Message{From: 2, To: 2, Type: pb.MsgHup})
> ```

We've previously spoken about checking whether we have support from a quorum of peers when campaigning. Would the current structure, where we withdraw support from all peers, campaign, and then grant support from all peers work well with that change?

This seems like a case where we'd want a more general implementation of store liveness. Would we be better served if we introduced that now?
___
*[`pkg/raft/raft_test.go` line 4341 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O8w6r9F8U_9LReX-Ele:-O8w6r9F8U_9LReX-Elf:bpuyuva) ([raw file](https://github.com/cockroachdb/cockroach/blob/3017a723a1fd9b0caac216dc7c460a5591b26770/pkg/raft/raft_test.go#L4341)):*
> ```Go
> 	}
> 
> 	if mock, ok := p.(*raft).storeLiveness.(*mockStoreLiveness); ok {
> ```

I like this pattern!


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> Reviewed 2 of 2 files at r7, 1 of 1 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132242)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/raft/raft_test.go` line 4165 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O8w1JC3703HUVewMK6w:-O9pAzFv3Ln22hDM1tSE:b-dg93pr) ([raw file](https://github.com/cockroachdb/cockroach/blob/96035db1e59585afec494eae70c27df7d2bc9d18/pkg/raft/raft_test.go#L4165)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nits:

s/simple//g

"It initially treats.."

"bump the supported epoch"
</blockquote></details>

Did you mean to still address these?
___
*[`pkg/raft/raft_test.go` line 4167 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O9pAgir9uu6_31V_ETR:-O9pAgir9uu6_31V_ETS:b-mujh3s) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad78e0484fd70ac2f406a26d6b0b0acb676de75e/pkg/raft/raft_test.go#L4167)):*
> ```Go
> // livenessEntry is an entry in the mockStoreLiveness state.
> type livenessEntry struct {
> 	// Controls whether supportFor returns true or false.
> ```

nit: we have a soft recommendation for starting comments with the field name. So something like: "isSupporting controls whether..."

Here and below.
___
*[`pkg/raft/raft_test.go` line 4191 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O9pB2qoD4daJr-Kmn0b:-O9pB2qoD4daJr-Kmn0c:b-obn0r3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad78e0484fd70ac2f406a26d6b0b0acb676de75e/pkg/raft/raft_test.go#L4191)):*
> ```Go
> // as live, but it can be configured to withdraw support, grant support, and
> // bump the support epoch to/from any two nodes.
> // Each node's state can independently be altered. This makes it possible to
> ```

nit: new line before the paragraph.

"can be altered independently" 
___
*[`pkg/raft/raft_test.go` line 4192 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O9pBBp3BWXY8GwE43me:-O9pBBp3BWXY8GwE43mf:b-2fmwib) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad78e0484fd70ac2f406a26d6b0b0acb676de75e/pkg/raft/raft_test.go#L4192)):*
> ```Go
> // bump the support epoch to/from any two nodes.
> // Each node's state can independently be altered. This makes it possible to
> // not have a shared liveness fabric between all the nodes in the test. This is
> ```

It also makes it possible to construct a uni-directional partition, if we want. 
___
*[`pkg/raft/raft_test.go` line 4213 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O9pByyjCY-ligGr7Fl7:-O9pByyjCY-ligGr7Fl8:b-a3qw0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad78e0484fd70ac2f406a26d6b0b0acb676de75e/pkg/raft/raft_test.go#L4213)):*
> ```Go
> // createStateEntryIfNotExist creates a new state entry for the given node if
> // it doesn't exist already.
> func (m mockStoreLiveness) createStateEntryIfNotExist(id pb.PeerID) {
> ```

s/createStateEntryIfNotExist/maybeCreateStoreLivenessEntry/g
___
*[`pkg/raft/raft_test.go` line 4270 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O9pCwpz9lQKOOmFa43C:-O9pCwpz9lQKOOmFa43D:b-491043) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad78e0484fd70ac2f406a26d6b0b0acb676de75e/pkg/raft/raft_test.go#L4270)):*
> ```Go
> 	m.createStateEntryIfNotExist(id)
> 	entry := m.state[id]
> 	entry.supportFromEpoch++
> ```

With this (and other) methods, we're not trying to keep the supported/supporter node in sync. That places the burden of keeping them in sync on the test writer, which makes it easy to make mistakes. 

I wonder if we should have common case wrappers for transitions we'll use frequently. For example, the case where we're bumping the `SupportFrom` epoch, we also want to have the `SupportFor` on the supporting node return the same epoch. Wdyt? 
___
*[`pkg/raft/raft_test.go` line 4165 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O9uDcVl5KofcaX1wBwb:-O9uDcVl5KofcaX1wBwc:bziic0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b266cf78b231538ecfe9984e26493a66b834367/pkg/raft/raft_test.go#L4165)):*
> ```Go
> }
> 
> // livenessEntry is an entry in the mockStoreLiveness state.
> ```

Let's move this to a new file.
___
*[`pkg/raft/raft_test.go` line 4419 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O9uDr3K77t8Ngqc5aGz:-O9uDr3K77t8Ngqc5aH-:b-4w6ixt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c844b87444a85995c6945fd934e746605dc710f0/pkg/raft/raft_test.go#L4419)):*
> ```Go
> // allWithdrawSupportForAndFromPeer makes all nodes withdraw support for and
> // from the given peer.
> func (nw *network) allWithdrawSupportForAndFromPeer(id pb.PeerID) {
> ```

The naming is a bit off here -- maybe let's call it `withdrawSupportForAndFromAllPeers`? 

(might be superseded by the comment to tie withdrawing support from/for in a single function)


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132242)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/raft/raft_test.go` line 4170 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O8qgDhrD0kN0bXWotvv:-O9v-T3A2vv3C1mzt72U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5d66fa455896d97e2d873b16847913f06199719/pkg/raft/raft_test.go#L4170)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

> For the tests that we need to change, I prepared this [document to list them all](https://docs.google.com/document/d/1cSbrdyF_3Tq_ZRsW-UTvxwRB4VylHXL2xEUPovkf3QA/edit).

Nice, thanks for doing this!

> Regarding the tests that require defortification. Do we have such tests as unit tests? Or do you mean that we will have them in the future?

I think tests that require the leader to step down because of `CheckQuorum` (from your doc) will fall into this category as well. The leader will not step down because of `CheckQuorum` until its `MaxLeadSupportUntil` is in the past. With `hlc.MaxTimestamp`, that'll never be true.

We can continue this discussion in the pod meeting later today.
</blockquote></details>

Done.
___
*[`pkg/raft/raft_test.go` line 4165 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O8w1JC3703HUVewMK6w:-O9vCNdd4eESPqcnmEl3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/96035db1e59585afec494eae70c27df7d2bc9d18/pkg/raft/raft_test.go#L4165)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Did you mean to still address these?
</blockquote></details>

Done.
___
*[`pkg/raft/raft_test.go` line 716 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O8w5gMM8iQDM1kauCsc:-O9v-1JdEvsfeG51W5HZ:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3017a723a1fd9b0caac216dc7c460a5591b26770/pkg/raft/raft_test.go#L716)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

We've previously spoken about checking whether we have support from a quorum of peers when campaigning. Would the current structure, where we withdraw support from all peers, campaign, and then grant support from all peers work well with that change?

This seems like a case where we'd want a more general implementation of store liveness. Would we be better served if we introduced that now?
</blockquote></details>

Done
___
*[`pkg/raft/raft_test.go` line 4341 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O8w6r9F8U_9LReX-Ele:-O9v-iBz9_qf4-wkAx9V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3017a723a1fd9b0caac216dc7c460a5591b26770/pkg/raft/raft_test.go#L4341)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I like this pattern!
</blockquote></details>

Done.
___
*[`pkg/raft/raft_test.go` line 4192 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O9pBBp3BWXY8GwE43me:-O9vCdRXCBAqVETR2BCX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad78e0484fd70ac2f406a26d6b0b0acb676de75e/pkg/raft/raft_test.go#L4192)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

It also makes it possible to construct a uni-directional partition, if we want.
</blockquote></details>

Done.
___
*[`pkg/raft/raft_test.go` line 4213 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O9pByyjCY-ligGr7Fl7:-O9vCl8O1hco0QS6BU01:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad78e0484fd70ac2f406a26d6b0b0acb676de75e/pkg/raft/raft_test.go#L4213)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

s/createStateEntryIfNotExist/maybeCreateStoreLivenessEntry/g
</blockquote></details>

Done.
___
*[`pkg/raft/raft_test.go` line 4270 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O9pCwpz9lQKOOmFa43C:-O9vKq2v7vq_To8lHrRo:bazh4on) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad78e0484fd70ac2f406a26d6b0b0acb676de75e/pkg/raft/raft_test.go#L4270)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

With this (and other) methods, we're not trying to keep the supported/supporter node in sync. That places the burden of keeping them in sync on the test writer, which makes it easy to make mistakes. 

I wonder if we should have common case wrappers for transitions we'll use frequently. For example, the case where we're bumping the `SupportFrom` epoch, we also want to have the `SupportFor` on the supporting node return the same epoch. Wdyt?
</blockquote></details>

I don't think it's feasible to do it from this layer given that there is no shared fabric.
___
*[`pkg/raft/raft_test.go` line 4165 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O9uDcVl5KofcaX1wBwb:-O9vKm_u5ybt1vbsWmwJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b266cf78b231538ecfe9984e26493a66b834367/pkg/raft/raft_test.go#L4165)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Let's move this to a new file.
</blockquote></details>

Done.
___
*[`pkg/raft/raft_test.go` line 4419 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O9uDr3K77t8Ngqc5aGz:-O9vKpM63c2EM_EoEg9d:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c844b87444a85995c6945fd934e746605dc710f0/pkg/raft/raft_test.go#L4419)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

The naming is a bit off here -- maybe let's call it `withdrawSupportForAndFromAllPeers`? 

(might be superseded by the comment to tie withdrawing support from/for in a single function)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> Reviewed 2 of 4 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132242)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/raft/raft_test.go` line 4270 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O9pCwpz9lQKOOmFa43C:-O9zotG76OEtf-IOwbum:b-q7tpjd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad78e0484fd70ac2f406a26d6b0b0acb676de75e/pkg/raft/raft_test.go#L4270)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

I don't think it's feasible to do it from this layer given that there is no shared fabric.
</blockquote></details>

Does it make sense to create a shared fabric? We'll still need these methods, to drive actions for store liveness entries, but we can have it such that tests interact with the shared fabric for the most part.


<!-- Sent from Reviewable.io -->

iskettaneh committed:
raft: test TestCannotCommitWithoutNewTermEntry w/ & wo/ store liveness

This commit changes TestCannotCommitWithoutNewTermEntry to allow testing
it when store liveness is both enabled and disabled.

References: #132241

Release note: None

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132242)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @iskettaneh)

___
*[`pkg/raft/raft_test.go` line 4270 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-O9pCwpz9lQKOOmFa43C:-OA5S-kx9aG2am7xHCXc:bbqu8ri) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad78e0484fd70ac2f406a26d6b0b0acb676de75e/pkg/raft/raft_test.go#L4270)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Does it make sense to create a shared fabric? We'll still need these methods, to drive actions for store liveness entries, but we can have it such that tests interact with the shared fabric for the most part.
</blockquote></details>

Discussed offline. I updated the mock implementation based on our discussion


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
I think this should be good to go after this round of comments. 

> Reviewed 2 of 2 files at r13, 4 of 4 files at r14, 1 of 1 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-:-OAIc_Ip7HN79W4TfQuS:b-zbosk3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 75 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIULvj9DXxa-bWc2xT:-OAIULvj9DXxa-bWc2xU:bid3sb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L75)):*
> ```Go
> 
> // SupportFrom implements the StoreLiveness interface.
> func (m MockStoreLiveness) SupportFrom(id pb.PeerID) (pb.Epoch, hlc.Timestamp) {
> ```

Goland is complaining that some of the methods have pointer receivers and others have value receivers. Let's convert all of these to pointer receivers? 

You'll need to make this change above when we do that:
```golang
var _ StoreLiveness = &MockStoreLiveness{}
```
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 133 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAITS-u7FXCEjtCYEz4:-OAITS-u7FXCEjtCYEz5:b-vrc7bj) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L133)):*
> ```Go
> // grantSupportFrom grants support from the given peer.
> func (m *MockStoreLiveness) grantSupportFrom(id pb.PeerID) {
> 	m.maybeCreateStoreLivenessEntry(id)
> ```

Do we still need these `maybeCreateStoreLivenessEntry` calls for each of these methods? Or can we get rid of them (and assert that an entry exists?)

```golang 
entry, ok := m.state[id]  
if !ok {  
  panic("...")  
}
```
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 140 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIV2P7-ltPz_INpNqP:-OAIV2P7-ltPz_INpNqQ:b-uia2y4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L140)):*
> ```Go
> 
> // WithdrawSupportFor withdraws support for the given peer.
> func (m *MockStoreLiveness) WithdrawSupportFor(id pb.PeerID) {
> ```

Does this need to be public?
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 177 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIUfZR4qG3MvmP55gu:-OAIUfZR4qG3MvmP55gv:bbi07tc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L177)):*
> ```Go
> 
> // MaybeAddPeer adds a peer to the liveness fabric if it doesn't already exist.
> // Also, it iterates over all existing peers and adds the new peer to their
> ```

nit: this "Also..." comment is an implementation detail which should be closer to the code and not in the top level function. I see you already have that below, so let's just drop this.
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 179 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIVWxEEY5nM54PcAiS:-OAIVWxEEY5nM54PcAiT:b-d8jrvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L179)):*
> ```Go
> // Also, it iterates over all existing peers and adds the new peer to their
> // state.
> func (l *LivenessFabric) MaybeAddPeer(id pb.PeerID) {
> ```

Related to my comment below about not needing `MaybeAddPeer` from other method calls in this file -- can we drop the "maybe", and instead have this be `AddPeer`?
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 197 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIYifl3Zd8-RhlRC7z:-OAIYifl3Zd8-RhlRC8-:b-3pygag) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L197)):*
> ```Go
> }
> 
> // BumpEpoch bidirectionally bumps the support epoch from `from` to `to`.
> ```

nit: bidirectionally is a bit confusing here -- consider rewording this to talk about the state transition here instead. Something like:

```
// BumpEpoch bumps the epoch supported by "from" for "to" and starts supporting the new epoch. We also update state on "to" to reflect support at this new epoch"
```
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 198 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIZm8mEqQPCzjpzOPA:-OAIZm8mEqQPCzjpzOPB:b-s1hep2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L198)):*
> ```Go
> 
> // BumpEpoch bidirectionally bumps the support epoch from `from` to `to`.
> func (l *LivenessFabric) BumpEpoch(from pb.PeerID, to pb.PeerID) {
> ```

nit: here, and elsehwere, instead of "from" and "to", should these be "from" and "for"? 
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 199 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIVLHU0c5jkei_RMu-:-OAIVLHU0c5jkei_RMu0:b-e67fov) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L199)):*
> ```Go
> // BumpEpoch bidirectionally bumps the support epoch from `from` to `to`.
> func (l *LivenessFabric) BumpEpoch(from pb.PeerID, to pb.PeerID) {
> 	l.MaybeAddPeer(from)
> ```

Do we need these calls to `MaybeAddPeer`? Or can we have the test initialize the `StoreLiveness` state correctly instead?
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 201 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAI_7nx2sYUGtztVctm:-OAI_7nx2sYUGtztVctn:b-4cyfn3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L201)):*
> ```Go
> 	l.MaybeAddPeer(from)
> 	l.state[from].bumpSupportForEpoch(to)
> ```

Should we be setting `IsSupported` to true here explicitly as well? (my comment suggestion assumes this).
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 206 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAI_NaR1TwB5PmCzLN5:-OAI_NaR1TwB5PmCzLN6:bzovr4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L206)):*
> ```Go
> }
> 
> // WithdrawSupport bidirectionally withdraws support from `from` to `to`.
> ```

nit: Same comment about bidirectionally -- let's talk about the state transition instead.

"WithdrawSupport withdraws support from `from` for `to`. We also update state on `to` to reflect withdrawal of support.."
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 215 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAI_ihn05fQNeRBdmjk:-OAI_iho-RmgEisyXnZ_:b-w2du5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L215)):*
> ```Go
> }
> 
> // WithdrawSupport bidirectionally grants support `from` from to `to`.
> ```

nit: same comment about bidirectional as above.
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 225 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAI_rteDfcYYvzzRQER:-OAI_rteDfcYYvzzRQES:b-xxarq7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L225)):*
> ```Go
> 
> // SetSupportExpired explicitly controls what SupportExpired returns regardless
> // of the timestamp.
> ```

"timestamp supplied to it".
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 231 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIazrb5GVQMtx3AYCl:-OAIazrb5GVQMtx3AYCm:bp72u33) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L231)):*
> ```Go
> }
> 
> // WithdrawAllSupportForPeer causes all peers in the liveness fabric to withdraw
> ```

nit: "WithdrawSupportForPeerFromAllPeers withdraws support for the target peer from all peers in the liveness fabric"
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 233 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIafWj9zUwRxuRb8dQ:-OAIafWj9zUwRxuRb8dR:be7upru) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L233)):*
> ```Go
> // WithdrawAllSupportForPeer causes all peers in the liveness fabric to withdraw
> // their support to the target peer.
> func (l *LivenessFabric) WithdrawAllSupportForPeer(target pb.PeerID) {
> ```

nit: "WithdrawSupportForPeerFromAllPeers"
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 240 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIbB3T5-fQAw0IWss_:-OAIbB3T5-fQAw0IWssa:b-1urq8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L240)):*
> ```Go
> 
> // GrantAllSupportForPeer causes all peers in the liveness fabric to grant
> // their support to the target peer.
> ```

nit: "GrantSupportForPeerFromAllPeers grants support for the target peer from all peers in the liveness fabric."
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 241 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIa5UmE0dSqxsOjPxO:-OAIa5UmE0dSqxsOjPxP:bv2knvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L241)):*
> ```Go
> // GrantAllSupportForPeer causes all peers in the liveness fabric to grant
> // their support to the target peer.
> func (l *LivenessFabric) GrantAllSupportForPeer(targetID pb.PeerID) {
> ```

nit: "GrantSupportForPeerFromAllPeers
___
*[`pkg/raft/raft_test.go` line 717 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIcRJv1j4xZFZnE1Ky:-OAIcRJv1j4xZFZnE1Kz:b-4h5bv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/866e3b5f139a1439cc550df9ee3f8b93f037990a/pkg/raft/raft_test.go#L717)):*
> ```Go
> 
> 		// Bumping all epochs will make all followers stop supporting the current
> 		// fortified leader.
> ```

Do we still need this? Or does the call to `WithdrawAllSupportForPeer(1)` give us this already?


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132242)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 75 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIULvj9DXxa-bWc2xT:-OAJ-iK_97EhU-AbGNw1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L75)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Goland is complaining that some of the methods have pointer receivers and others have value receivers. Let's convert all of these to pointer receivers? 

You'll need to make this change above when we do that:
```golang
var _ StoreLiveness = &MockStoreLiveness{}
```
</blockquote></details>

Done.
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 133 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAITS-u7FXCEjtCYEz4:-OAJ2wZU680Q-aJ2cykx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L133)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Do we still need these `maybeCreateStoreLivenessEntry` calls for each of these methods? Or can we get rid of them (and assert that an entry exists?)

```golang 
entry, ok := m.state[id]  
if !ok {  
  panic("...")  
}
```
</blockquote></details>

Done.
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 140 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIV2P7-ltPz_INpNqP:-OAJ2xsQ3_O5n_pLHwKb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L140)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Does this need to be public?
</blockquote></details>

Done.
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 179 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIVWxEEY5nM54PcAiS:-OAJ2zzB343rCckD6gdV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L179)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Related to my comment below about not needing `MaybeAddPeer` from other method calls in this file -- can we drop the "maybe", and instead have this be `AddPeer`?
</blockquote></details>

Done.
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 199 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIVLHU0c5jkei_RMu-:-OAJ7-LD2Q-vDZYH-L4p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L199)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Do we need these calls to `MaybeAddPeer`? Or can we have the test initialize the `StoreLiveness` state correctly instead?
</blockquote></details>

Done.
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 201 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAI_7nx2sYUGtztVctm:-OAJ7qVuDPZx0zNE9-Wq:b4ndsa9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L201)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we be setting `IsSupported` to true here explicitly as well? (my comment suggestion assumes this).
</blockquote></details>

I did it for now, but I am not entirely sure about it. My thought that we might need to bump the epoch without granting support to simulate something
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 225 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAI_rteDfcYYvzzRQER:-OAJ85k35SQRcx0y6lYq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L225)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

"timestamp supplied to it".
</blockquote></details>

Done.
___
*[`pkg/raft/raft_test.go` line 717 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIcRJv1j4xZFZnE1Ky:-OAJH2dyDUHCGXletc5h:b1kl480) ([raw file](https://github.com/cockroachdb/cockroach/blob/866e3b5f139a1439cc550df9ee3f8b93f037990a/pkg/raft/raft_test.go#L717)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Do we still need this? Or does the call to `WithdrawAllSupportForPeer(1)` give us this already?
</blockquote></details>

Now, BumpAllSupportEpochs() will restore support (but at a higher epoch). 
So the workflow is this:

1) Withdraw support
2) Bump epoch (and implicitly restore support)

We need (1) to avoid the leader refortifying the follower at the higher epoch (in 2). Technically though, we can just do (1)


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo comments.

> Reviewed 5 of 5 files at r16, 4 of 4 files at r17, 1 of 1 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-:-OAJKa9z6aUsPhpBi37K:b46bmkf)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 201 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAI_7nx2sYUGtztVctm:-OAJIsLcDsvZ4qsQB1s0:bjo6idp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6319fe453773f45ce7773bb4f2157d2e2c38a666/pkg/raft/raftstoreliveness/mock_store_liveness.go#L201)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

I did it for now, but I am not entirely sure about it. My thought that we might need to bump the epoch without granting support to simulate something
</blockquote></details>

I think you can still do that with bump + withdraw. You can do things the other way around too, so the question really is which default makes sense -- I think this one does. 
___
*[`pkg/raft/raftstoreliveness/mock_store_liveness.go` line 253 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAJJ8ulEIFtfyyukrty:-OAJJ8ulEIFtfyyukrtz:bgx81bb) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcedf42d30b3bfc73484e92ca3337d8772b5a492/pkg/raft/raftstoreliveness/mock_store_liveness.go#L253)):*
> ```Go
> }
> 
> // WithdrawSupport withdraws the support by "fromID" for "forID". We also update
> ```

nit: now that we've renamed these, you could get rid of the quotes around "fromID" and "forID" -- typically, we don't wrap parameter names in quotes.

(here and elsewhere). 
___
*[`pkg/raft/raft_test.go` line 717 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/132242#-OAIcRJv1j4xZFZnE1Ky:-OAJKGUZ44LiWeVPobCE:bjmg2b0) ([raw file](https://github.com/cockroachdb/cockroach/blob/866e3b5f139a1439cc550df9ee3f8b93f037990a/pkg/raft/raft_test.go#L717)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Now, BumpAllSupportEpochs() will restore support (but at a higher epoch). 
So the workflow is this:

1) Withdraw support
2) Bump epoch (and implicitly restore support)

We need (1) to avoid the leader refortifying the follower at the higher epoch (in 2). Technically though, we can just do (1)
</blockquote></details>

Ack. I think we should just do 1 and not do 2. Like we spoke about offline, doing 2 can lead to a race where the old leader is able to re-fortify.


<!-- Sent from Reviewable.io -->

iskettaneh commented:
bors r+

arulajmani commented:
blathers backport 24.3
